### ISSUE TITLE
DOMParser Api not working as expected when <nav> tag is a child of <p> tag.

### ISSUE THREAD

**Issue by balajee500** [mdn/yari#4304]
I am trying to traverse through the node based on input HTML string. Basically. _In the below HTML String, the **nav** tag is a child of **p** tag,_

```
<div` class="slds-media__body" data-aura-rendered-by="10:1819;a">
	<p class="entity-label slds-line-height--reset" data-aura-rendered-by="11:1819;a">
		<nav role="navigation" aria-label="Breadcrumbs" data-aura-rendered-by="12:1819;a">
			<ol class="slds-breadcrumb slds-list--horizontal slds-wrap" data-aura-rendered-by="13:1819;a">
				<li class="slds-breadcrumb__item entityName" data-aura-rendered-by="14:1819;a">
					<span class="slds-p-right--x-small entityName uiOutputText" data-aura-rendered-by="17:1819;a" data-aura-class="uiOutputText">Dashboards</span>
				</li>
			</ol>
		</nav>
	</p>
</div>
```
When the HTML String is parsed into DOM using DOMParser API(new DOMParser().parseFromString(htmlContent, "text/html");), _the **nav** tag placed adjacent to **P** tag.  which is not same as the HTML input String_. Refer the below output.

```
<html>
	  <head></head>
	  <body>
		<div class="slds-media__body" data-aura-rendered-by="10:1819;a">
		  <p class="entity-label slds-line-height--reset" data-aura-rendered-by="11:1819;a"></p>
		  <nav role="navigation" aria-label="Breadcrumbs" data-aura-rendered-by="12:1819;a">
			<ol class="slds-breadcrumb slds-list--horizontal slds-wrap" data-aura-rendered-by="13:1819;a">
			  <li class="slds-breadcrumb__item entityName" data-aura-rendered-by="14:1819;a">
				<span class="slds-p-right--x-small entityName uiOutputText" data-aura-rendered-by="17:1819;a" data-aura-class="uiOutputText">Dashboards</span>
			  </li>
			</ol>
		  </nav>
		  <p></p>
		</div>
	  </body>
	</html>
```

Also refer the below screenshot,
![image](https://user-images.githubusercontent.com/36754725/126903872-b9be1fd6-7bb9-48bc-824f-d7888f501230.png)

Run the below script to reproduce the issue.

```
var htmlContent = "<div class=\"slds-media__body\" data-aura-rendered-by=\"10:1819;a\"><p class=\"entity-label slds-line-height--reset\" data-aura-rendered-by=\"11:1819;a\"><nav role=\"navigation\" aria-label=\"Breadcrumbs\" data-aura-rendered-by=\"12:1819;a\"><ol class=\"slds-breadcrumb slds-list--horizontal slds-wrap\" data-aura-rendered-by=\"13:1819;a\"><li class=\"slds-breadcrumb__item entityName\" data-aura-rendered-by=\"14:1819;a\"><span class=\"slds-p-right--x-small entityName uiOutputText\" data-aura-rendered-by=\"17:1819;a\" data-aura-class=\"uiOutputText\">Dashboards</span></li></ol></nav></p></div>";
new DOMParser().parseFromString(htmlContent, "text/html");
```

The above issue causing application to behave inappropriate. Please help me with the above issue!!

**Comment by tannerdolby** [comment]
Before addressing the parser API stuff. In terms of HTML semantics, why exactly would you place a navigation element `<nav>` inside a `<p>` tag?

**Comment by tannerdolby** [comment]
Ok so there is actually a lot going wrong here semantically speaking. Your placing a `<nav>` element inside a `<p>` tag which is invalid HTML. The paragraph element [`<p>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/p) only has permitted content of [phrasing content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#phrasing_content) which `<nav>` is not phrasing content its [flow content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#flow_content). 

> Paragraphs are block-level elements, and notably will automatically close if another block-level element is parsed before the closing </p> tag. See “Tag omission” below.

In conclusion, not only are you you feeding the DOMParser API invalid HTML but when a block-level element like `<p>` has another block-level element nested inside it (like you've done with `<p><nav></nav></p>`) then the `<p>` tag will be automatically closed when parsed. Hopefully this clears things up for you! This isn't an error with the API, its actually the expected behavior described in the [Spec](https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#dom-parsing-and-serialization).

To prevent this from happening, replace your `<p>` tag with a `<div>` element if your only going to have one nested child `<nav>` element inside that container. This way your HTML DOM structure won't be changed from the `<p>` tags being automatically closed when parsed and instead the expected HTML will be parsed by the API. I threw together a CodePen example to better illustrate the behavior if you want to have a look there. 

This should help! @balajee500

**Comment by tannerdolby** [comment]
Although this conversation about DOMParser API is not explicitly `mdn/yari` related, I think its worth potentially adding a small bit to the [DOMParser API](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) docs illustrating that this situation will happen for some elements. Which MDN docs does state on the paragraph element page that `<p>` tags will be automatically closed if another block level element is parsed before the closing `</p>` tag, but I think adding a note within the DOMParser API to warn people about these scenarios would be helpful. The numbered list are things that I think are important to think about when parsing with DOMParser API, as when parsing block-level elements with nested

1. Using valid and semantic HTML when passing HTML/XML etc to DOMParser with `parseFromString`.
2. Note about [Tag omission]() and the type (block / inline) of that element being parsed to inform users that they must be careful when nesting certain block level elements inside other block-level elements as in certain situations the tags will be automatically closed when parsed due to Tag Omission.

cc @peterbe

**Comment by balajee500** [comment]
@tannerdolby 
In your [CodePen](https://codepen.io/tannerdolby/pen/NWjXjxY?editors=1011) example, there is a block level element ```<nav>``` has another block level element ```<ol>``` nested inside it```(like <nav><ol></ol></nav>)```. As per your above comment, the nav tag should be automatically closed when parsed. but that does not happen. so, can you clear things up for me?

**Comment by tannerdolby** [comment]
> In your CodePen example, there is a block level element `<nav>` has another block level element `<ol>` nested inside it(like `<nav><ol></ol></nav>`). As per your above comment, the nav tag should be automatically closed when parsed. but that does not happen. so, can you clear things up for me?

Yeah sure thing. So this "automatically" closing case related to the DOMParser API seems to be strictly related to using `<p>` tags for nesting html content other than text (as paragraph tags aren't meant to act as a container like your attempting in your HTML), and the fact that the `<nav>` is a block level element means the `<p>` tag doesn't need a closing tag from "Tag omission" but regardless of closing the `<p>` tag or not the parser will automatically close it. As its not semantically "right" to nest anything in a `<p>` tag other than phrasing content, it would be better in your case to just replace the paragraph element with a `<div>` as this "case" was probably imagined when the API was created (could be a bug) but I think its expected behavior as your using `<p>` for something its not meant for.

So the reason why the `<nav>` tag isn't automatically closed when parsed even though it has another block level element nested, is because the parser is "OK" with this html structure as `<ol>` is indeed flow content which is valid. The [`<nav>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/nav) has permitted content of flow content so having the nested `<ol>` is OK in terms of valid HTML which I'm sure the parser takes into account with other HTML standard rules. Therefore the `<nav>` tag is not automatically closed and the "nested block level" case is really only applying to the `<p>` tag with nested block level content (which is not its permitted content). The parser closing the `<p>` tag has to do specifically with your usage of the `<p>` element. When working with `marked.js` (library for parsing markdown) I've seen some weird stuff happen when parsing but honestly this behavior of closing the `<p>` tag seems fine to me as its not semantically correct to nest a `<nav>` menu within a `<p>` tag, from a screen reader standpoint and from a HTML semantics view. 

This is indeed an odd situation and I hope my explanations have helped a bit to better understand why the `<p>` tag is being closed but I'm glad you documented it so we can dig deeper and hopefully provide more information for users that encounter this issue as well.
